File: code/library/src/main/java/org/fbme/lib/iec61499/parser/ConverterArguments.java
package org.fbme.lib.iec61499.parser;

import org.fbme.lib.iec61499.IEC61499Factory;
import org.fbme.lib.st.STFactory;
import org.jdom.Element;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface ConverterArguments {

    @Nullable
    Element getElement();

    @NotNull
    IEC61499Factory getFactory();

    @NotNull
    STFactory getStFactory();

    @NotNull
    IdentifierLocus getIdentifierLocus();

    default ConverterArgumentsHolder with(Element element) {
        return new ConverterArgumentsHolder(getFactory(), getStFactory(), getIdentifierLocus(), element);
    }

    static ConverterArguments create(IEC61499Factory factory, STFactory stFactory, IdentifierLocus locus, Element element) {
        return new ConverterArgumentsHolder(factory, stFactory, locus, element);
    }

}

File: code/library/src/main/kotlin/org/fbme/lib/iec61499/parser/ConverterArguments.kt
package org.fbme.lib.iec61499.parser

import org.fbme.lib.iec61499.IEC61499Factory
import org.fbme.lib.st.STFactory
import org.jdom.Element

interface ConverterArguments {
    val element: Element?
    val factory: IEC61499Factory
    val stFactory: STFactory
    val identifierLocus: IdentifierLocus

    fun with(element: Element?): ConverterArgumentsHolder {
        return ConverterArgumentsHolder(factory, stFactory, identifierLocus, element)
    }

    companion object {
        fun create(
            factory: IEC61499Factory,
            stFactory: STFactory,
            locus: IdentifierLocus,
            element: Element?
        ): ConverterArguments {
            return ConverterArgumentsHolder(factory, stFactory, locus, element)
        }
    }
}
